dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityGeoValidationHook.java

```diff
@@ -0,0 +1,81 @@
+package org.hisp.dhis.tracker.validation.hooks;
+
+/*
+ * Copyright (c) 2004-2020, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.organisationunit.FeatureType;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.springframework.stereotype.Component;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class TrackedEntityGeoValidationHook
+    extends AbstractTrackerValidationHook
+{
+
+    @Override
+    public int getOrder()
+    {
+        return 4;
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerBundle bundle )
+    {
+        if ( bundle.getImportStrategy().isDelete() )
+        {
+            return Collections.emptyList();
+        }
+
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, this.getClass() );
+
+        for ( TrackedEntity trackedEntity : bundle.getTrackedEntities() )
```

Comment:
seems we are iterating the TEIs every time we want to validate something. Not sure if this is a good idea.

For me, it makes much more sense if just have one iteration and make calls to various validation hooks within this iteration. We could also somehow put the validations in an orderly fashion and a failure on top means no need to validate lower ones - we just return. 

For example, if user has no write access to the orgunit under the TEI, then no need to check the attributes - reject the whole thing.

Reply:
@abyot I totally understand what you are saying here, this was my initial reaction too. However after refactoring the old code into this new style I think it has some benefits, first a lot of the validations can be run even if some fails, that can mean in practice that the importer will get a list of more errors than just the first failing one, this will probably be more user friendly. Also that being said, there is an option to chose a "fail fast" import. The validation system will then throw an exception at the first error and only return the first error found. When it comes to reiterate the objects on each hook I'm not sure how big the performance impact actually will be compared to running all the hooks in a loop on each "row". I suspect that this actually has not so big impact relatively to the more expensive operations done, but I could be wrong. I think changing the semantics to this can be done relatively easily later on if we suspect there is performance issues.

Reply:
yes, true @netroms, my main concern is the performance we will be paying with the current approach. I am afraid of ending up in the same situation of the tracker import being too slow ... we reached to this situation partly because we were looking after functionalities.

I can see for example some metadatas being fetched in each of these iterations... org unit and program are heavy. yes there is caching,  but still... not sure on the hit rate especially when importing huge payload. I am not sure also worse would be if we are checking TEI access in these iterations.